Overload HLG values method

jrbourbeau opened the pull request:
This PR overloads the built-in `Mapping.values()` method for `HighLevelGraph`s. The current `values()`  method makes calls to the graph `__getitem__()` method which, for large-ish graphs, can be slow (see https://github.com/dask/distributed/blob/d202e6253ed8ddc7919d0d4f128d88954e9859b8/distributed/client.py#L4336 for a relevant use case).

The `values()` method implemented in this PR instead concatenates the values for each HLG layer, avoiding repeated calls to the `HighLevelGraph.__getitem__()` method.

Timing the following example

```python
In [1]: import dask.array as da 
   ...:  
   ...: a = da.random.randint(0, 10, size=10_000, chunks=10) 
   ...:  
   ...: # Create more layers for HighLevelGraph 
   ...: for i in range(10): 
   ...:     a = a + 1 
   ...:                                                                                                                     
```

on the current `master` branch yields

```python
In [2]: %timeit -n 5 list(a.dask.values())                                                                                  
81.8 ms ± 4.44 ms per loop (mean ± std. dev. of 7 runs, 5 loops each)
```

while with the changes in this PR we get a significant speedup factor

```python
In [2]: %timeit -n 5 list(a.dask.values())                                                                                  
5.6 ms ± 2.68 ms per loop (mean ± std. dev. of 7 runs, 5 loops each)
```



- [x] Tests added / passed
- [x] Passes `flake8 dask`

jcrist commented on dask/highlevelgraph.py:
```diff
@@ -153,6 +153,9 @@ def items(self):
         return toolz.unique(toolz.concat(self.layers.values()))
 
+    def values(self):
+        return toolz.concat(layer.values() for layer in self.layers.values())
```
This seems fine to me in practice, but does create a structure different than a `dict_values` view, which means that it has different properties. We should at least return a list instead of a generator, but I'm not sure if there's a better structure to use.

jcrist replied:
We should probably also do this for `keys` and `items`?

mrocklin replied:
Perhaps we can just implement `items` and have the others come for free?

Why a list rather than a generator?

jcrist replied:
Because a generator doesn't fullfill the contract of `values`/`keys`/items` (e.g. `item in d.values()`). That's why Python 2 also had `itervalues`/`iterkeys`, `iteritems`.

jrbourbeau replied:
@jcrist am I missing something? It looks like `item in d.values()` is fullfilled

```python
In [1]: import dask.array as da                                                                                             

In [2]: a = da.ones(10, chunks=2)                                                                                           

In [3]: hlg = a.dask                                                                                                        

In [4]: list(hlg.values())[0] in hlg.values()                                                                               
Out[4]: True
```

jrbourbeau replied:
@mrocklin to your point, the current `items` method already avoids the `HighLevelGraph.__getitem__()` method, so could be used as is to build the `keys` and `values` logic. E.g. something like

```python
def values(self):
    for _, val in self.items():
        yield val
```

jcrist replied:
Looks like `itertools.chain` implements `__contains__`, but it's implemented by consuming the generator until a match is found, leading to poor performance and odd behavior:

```python
In [34]: res = toolz.concat([[1, 2, 3], [4], [5, 6, 7]])

In [35]: 1 in res
Out[35]: True

In [36]: 1 in res
Out[36]: False
```

jrbourbeau commented:
I updated `items` to return a concrete `list` instead of a generator, and implemented `keys` and `values` methods which use the `items` method. At this point, I think all the [`values`/`keys`/`items` properties](https://docs.python.org/3/library/stdtypes.html#dictionary-view-objects) should be satisfied (except being lazy).

Here are benchmark results using the [original example HLG](https://github.com/dask/dask/pull/4918#issue-287234784):

<details>
<summary>on the current master branch</summary>

```python
In [2]: %timeit -n 10 list(a.dask.keys())                                                                             
4.8 ms ± 1.31 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [3]: %timeit -n 10 list(a.dask.values())                                                                           
78.2 ms ± 295 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [4]: %timeit -n 10 list(a.dask.items())                                                                            
8.09 ms ± 300 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
</details>

<details>
<summary>with changes in this PR</summary>

```python
In [2]: %timeit -n 10 a.dask.keys()                                                                                         
7.58 ms ± 1.41 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [3]: %timeit -n 10 a.dask.values()                                                                                       
7.32 ms ± 592 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [4]: %timeit -n 10 a.dask.items()                                                                                        
6.57 ms ± 328 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
</details>

@jcrist when you get a chance to take a look at this, any additional comments are appreciated

jcrist commented:
Overall this looks good to me. Could you add a small test for `keys` and `items` (unless that's already somewhere else in the test file).

jcrist commented:
LGTM. Merging.
